feat: User can add Least Squares Line to scatterplot (PT-181939979)

emcelroy opened the pull request:
PT Story: https://www.pivotaltracker.com/story/show/181939979

emcelroy commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-component.tsx:
```diff
@@ -0,0 +1,359 @@
+
+      const equation = equationDiv.selectAll("p").filter((d, i) => i === lineIndex)
+      equation.call(
+        drag<any, any>().on("drag", (e) => handleMoveEquation(e, false, lineIndex))
```
I did try to find more specific types to use here, but had trouble and decided not to spend anymore time on it for now.

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:
```diff
@@ -0,0 +1,157 @@
+    const casesInPlot = dataConfig.subPlotCases(cellKey)
+    const caseValues: Point[] = []
+    casesInPlot.forEach((c: ICase) => {
+      const caseValueX = Number(dataset?.getValue(c.__id__, xAttrId))
+      const caseValueY = Number(dataset?.getValue(c.__id__, yAttrId))
```
```suggestion
      const caseValueX = dataset?.getNumeric(c.__id__, xAttrId)
      const caseValueY = dataset?.getNumeric(c.__id__, yAttrId)
```
You don't need to perform the numeric conversion yourself.

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:
```diff
@@ -0,0 +1,157 @@
+  confidenceValues(iX: number, caseValues: Point[], cellKey: Record<string, string>, lineIndex: number) {
+    const lines = self.lines.get(self.instanceKey(cellKey))
+    const line = lines?.[lineIndex]
+    const interceptIsLocked = false
```
Should there be a TODO comment here about supporting the locked case?

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:
```diff
@@ -0,0 +1,157 @@
+    const lines = self.lines.get(self.instanceKey(cellKey))
+    const line = lines?.[lineIndex]
+    const interceptIsLocked = false
+    const slopeIntercept = leastSquaresLinearRegression(caseValues, interceptIsLocked)
+    const { count, mse, xSumSquaredDeviations, xMean } = slopeIntercept
```
Seems like `slopeIntercept` is the wrong name here, and maybe you don't need an intermediate variable at all:
```suggestion
    const { count, mse, xSumSquaredDeviations, xMean } = leastSquaresLinearRegression(caseValues, interceptIsLocked)
```

kswenson commented on v3/src/components/graph/utilities/graph-utils.ts:
```diff
@@ -275,3 +293,220 @@ export function computeSlopeAndIntercept(xAxis?: IAxisModel, yAxis?: IAxisModel)
+  yMean?: number | null
+}
+
+export const leastSquaresLinearRegression = (iValues: any, iInterceptLocked: boolean) => {
```
Type?

kswenson commented on v3/src/components/graph/utilities/graph-utils.ts:
```diff
@@ -275,3 +293,220 @@ export function computeSlopeAndIntercept(xAxis?: IAxisModel, yAxis?: IAxisModel)
+  const foundIndex = t_quantile_at_0975_for_df.findIndex((iPair: any) => iPair[0] > iDf)
+  return foundIndex <= 0 ? 1.96 : t_quantile_at_0975_for_df[foundIndex - 1][1]
+}
+export interface ISlopeIntercept {
```
Perhaps `IRegression` or `IRegressionResults`?

kswenson commented on v3/src/components/graph/utilities/graph-utils.ts:
```diff
@@ -275,3 +293,220 @@ export function computeSlopeAndIntercept(xAxis?: IAxisModel, yAxis?: IAxisModel)
+  return foundIndex <= 0 ? 1.96 : t_quantile_at_0975_for_df[foundIndex - 1][1]
+}
+export interface ISlopeIntercept {
+  count: number | null
+  intercept: number | null
+  mse: number | null // mean squared error
+  rSquared: number | null
+  sdResiduals: number | null
+  slope: number | null
+  sse: number // sum of squared errors
+  sumSquaresResiduals?: number | null
+  xMean: number | null
+  xSumSquaredDeviations: number | null
+  yMean?: number | null
```
Is there a reason to use `null` here? My inclination would have been to make them all optional properties so they get initialized to `undefined` automatically.

kswenson commented on v3/src/components/graph/utilities/graph-utils.ts:
```diff
@@ -275,3 +293,220 @@ export function computeSlopeAndIntercept(xAxis?: IAxisModel, yAxis?: IAxisModel)
+}
+
+export const leastSquaresLinearRegression = (iValues: any, iInterceptLocked: boolean) => {
+  const tSlopeIntercept: ISlopeIntercept = {
```
Seems like there are better names.

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:
```diff
@@ -0,0 +1,157 @@
+    const { count, mse, xSumSquaredDeviations, xMean } = slopeIntercept
+    if (
+      !line ||
+      (!count && count !== 0) ||
+      (!mse && mse !== 0) ||
+      (!xMean && xMean !== 0) ||
+      (!xSumSquaredDeviations && xSumSquaredDeviations !== 0)
```
I don't understand these test pairs -- if it's meant to be a test for `null`/`undefined` then I'd just use `count != null`, etc.

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:
```diff
@@ -0,0 +1,157 @@
+          xAttrId, yAttrId, cellKey, dataConfig, instanceKey, false, legendCats[j]
+        )
+        if (
+          (intercept === null && intercept !== 0) ||
+          (rSquared === null && rSquared !== 0) ||
+          (slope === null && slope !== 0) ||
+          (sdResiduals === null && sdResiduals !== 0)
```
A value that `=== null` is by definition `!== 0`.

emcelroy replied:
Thanks. I still get confused sometimes by `null`, `0`, and `undefined` and which is considered equal to which. I need to get that squared away in my head.

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-registration.tsx:
```diff
@@ -0,0 +1,96 @@
+  const adornmentsStore = graphModel.adornmentsStore
+  const existingAdornment = adornmentsStore.findAdornmentOfType<ILSRLAdornmentModel>(kLSRLType)
+
+  const handleBoxPlotSetting = (checked: boolean) => {
+    const existingBoxPlotAdornment = adornmentsStore.findAdornmentOfType<ILSRLAdornmentModel>(kLSRLType)
```
Naming?

kswenson commented on v3/src/components/graph/utilities/graph-utils.ts:
```diff
@@ -275,3 +293,220 @@ export function computeSlopeAndIntercept(xAxis?: IAxisModel, yAxis?: IAxisModel)
+   * @returns {{count: {Number}, xSum: {Number}, xSumOfSquares: {Number}, xSumSquaredDeviations: { Number},
+  *          ySum: {Number}, ySumOfSquares: {Number}, ySumSquaredDeviations: {Number}, sumOfProductDiffs: {Number} }
+  */
+ const computeBivariateStats = (iCoordPairs: any) => {
```
Type?

kswenson commented on v3/src/components/graph/utilities/graph-utils.ts:
```diff
@@ -275,3 +293,220 @@ export function computeSlopeAndIntercept(xAxis?: IAxisModel, yAxis?: IAxisModel)
+    let tSumDiffsY = 0
+
+    // Under certain circumstances (adding new case) an empty value can sneak in here. Filter out.
+    iCoordPairs = iCoordPairs.filter((iPair: any) => {
+      return !(!iPair.x || !iPair.y)
+    })
+   iCoordPairs.forEach(function (iPair: any) {
```
Types?

kswenson commented on v3/src/utilities/translation/lang/en-US.json5:
Generally, we've been adding new v3 strings at the top of the file. Are these meant to be v2 strings as well? Has the working been vetted by the appropriate people? @bfinzer?

emcelroy replied:
@kswenson I lifted the new strings from [this V2 translation file](https://github.com/concord-consortium/codap/blob/master/lang/strings/en-US.json) in `master`. I believe they were added in V2 relatively recently which is why they weren't already in the V3 file. In that case, is it OK that they're located where they are in the V3 file? Or should they still be moved to the top?

kswenson replied:
No, it suggests that we should do a better job of keeping the v3 file up-to-date with changes in v2, so what you did was exactly correct.

kswenson reviewed (changes requested):
👍 Impressive work -- I made a bunch of suggestions. Some of them I realize in retrospect probably relate to code that was copied from v2, but I still think it's worth a pass to clean up some of the naming and typing issues because otherwise they'll probably never be addressed.

bfinzer commented:
Found bugs by checking out an running the branch. These could be logged in a separate story or stories if you prefer: 
* When the intercept is negative the "+" in the equation should not be present
* The slope should only be parenthesized if there are units present. Since we're not dealing with units yet, the parentheses will come in later.
* I rename the x-axis attribute to be a single letter; e.g. *x*. In that case we don't parenthesize it in the equation.
* The equation color should match the line color when there is a legend
* When dragging the equation, if the mouse goes down near one end, the equation "jumps" so the mouse is centered. It shouldn't jump.
* I put Top_Speed on x and Max_Height on y and showed LSRL. Then I added Type as legend. Finally, I added Inversions to the right axis. The lines did not show in the top subplot.

bfinzer reviewed (approved):
👍🏻LGTM
Kirk has commented on the code. My contribution at this stage is noting bugs. I notice that some of the bugs I found apply equally to movable line equation.

pjanik commented:
@emcelroy, in case you're surprised by my last commit - I just merged main into your PR, as we both touched the same files (plotted function component and model). But it shouldn't affect your work at all. I just adjusted the component to use graph-utils helpers that you added in this PR (and removed copies of the same code from the component file itself - my previous approach).

emcelroy commented:
> @emcelroy, in case you're surprised by my last commit - I just merged main into your PR, as we both touched the same files (plotted function component and model). But it shouldn't affect your work at all. I just adjusted the component to use graph-utils helpers that you added in this PR (and removed copies of the same code from the component file itself - my previous approach).

@pjanik Thanks for doing that and letting me know. Makes sense 👍

emcelroy commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-component.tsx:
```diff
@@ -0,0 +1,359 @@
+  const graphModel = useGraphContentModelContext()
+  const dataConfig = useGraphDataConfigurationContext()
+  const layout = useAxisLayoutContext()
+  const xScale = layout.getAxisScale("bottom") as ScaleNumericBaseType
+  const yScale = layout.getAxisScale("left") as ScaleNumericBaseType
+  const xAttrType = dataConfig?.attributeType("x")
+  const yAttrType = dataConfig?.attributeType("y")
+  const xSubAxesCount = layout.getAxisMultiScale("bottom")?.repetitions ?? 1
+  const ySubAxesCount = layout.getAxisMultiScale("left")?.repetitions ?? 1
+  const xCatSet = layout.getAxisMultiScale("bottom")?.categorySet
+  const xCats = xAttrType === "categorical" && xCatSet ? Array.from(xCatSet.values) : [""]
+  const yCatSet = layout.getAxisMultiScale("left")?.categorySet
+  const yCats = yAttrType === "categorical" && yCatSet ? Array.from(yCatSet.values) : [""]
+  const xCellCount = xCats.length * xSubAxesCount
+  const yCellCount = yCats.length * ySubAxesCount
+  const allAttributes = dataConfig?.dataset?.attributes
+  const xAttrId = dataConfig?.attributeID("x") || ""
+  const yAttrId = dataConfig?.attributeID("y") || ""
+  const xAttr = allAttributes?.find(attr => attr.id === xAttrId)
+  const yAttr = allAttributes?.find(attr => attr.id === yAttrId)
+  const xAttrName = xAttr?.name ?? ""
+  const yAttrName = yAttr?.name ?? ""
+  const instanceKey = model.instanceKey(cellKey)
+  const classFromKey = model.classNameFromKey(cellKey)
```
That makes sense, but is something I'd like to attack in a separate piece of work.

emcelroy commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-component.tsx:
```diff
@@ -0,0 +1,359 @@
+
+  // Refresh values on axis changes
+  useEffect(function refreshAxisChange() {
+    return mstAutorun(() => {
+      // We observe changes to the axis domains within the autorun by extracting them from the axes below.
+      // We do this instead of including domains in the useEffect dependency array to prevent domain changes
+      // from triggering a reinstall of the autorun.
+      const { domain: xDomain } = xAxis // eslint-disable-line @typescript-eslint/no-unused-vars
+      const { domain: yDomain } = yAxis // eslint-disable-line @typescript-eslint/no-unused-vars
+      updateLSRL()
+    }, { name: "LSRLAdornmentComponent.refreshAxisChange" }, model)
```
@kswenson I tried this but the line would then not dynamically update when you did things like drag a case dot around the plot. I'm not sure why exactly. I left it as is, for now at least. If changing it is important, though, I can try again.

kswenson replied:
I was thinking that the`mstAutorun` was specifically reacting to domain changes. Apparently, it's also reacting to stuff that's being accessed in `updateLSRL()`. Leave it for now and we can come back to it later if it becomes a problem, e.g. if it's calling `updateLSRL()` too frequently.

emcelroy commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-component.tsx:
```diff
@@ -0,0 +1,355 @@
+      lineObj.equation = equationDiv
+      lineObjectsRef.current = [...lineObjectsRef.current, lineObj]
+
+      const equation = equationDiv.select<HTMLElement>(`#lsrl-equation-${model.classNameFromKey(cellKey)}-${lineIndex}`)
+      equation?.call(
+        drag<HTMLElement, unknown>().on("drag", (e) => handleMoveEquation(e, false, lineIndex))
+          .on("end", (e) => handleMoveEquation(e, true, lineIndex))
+      )
+    }
```
In case it helps someone in the future, the key to being able to use `drag<HTMLElement, unknown>` on line 321 instead of resorting to `drag<any, any>` was adding `<HTMLElement>` to the select in line 319. Without the latter, there was an error thrown by line 321 `Type 'Selection<BaseType, unknown, HTMLElement, any>' is not assignable to type 'Selection<HTMLElement, unknown, any, any>'` `BaseType` could potentially equal `null` which isn't assignable to `HTMLElement`.

emcelroy commented:
> Found bugs by checking out an running the branch. These could be logged in a separate story or stories if you prefer:
> 
> * When the intercept is negative the "+" in the equation should not be present
> * The slope should only be parenthesized if there are units present. Since we're not dealing with units yet, the parentheses will come in later.
> * I rename the x-axis attribute to be a single letter; e.g. _x_. In that case we don't parenthesize it in the equation.
> * The equation color should match the line color when there is a legend
> * When dragging the equation, if the mouse goes down near one end, the equation "jumps" so the mouse is centered. It shouldn't jump.
> * I put Top_Speed on x and Max_Height on y and showed LSRL. Then I added Type as legend. Finally, I added Inversions to the right axis. The lines did not show in the top subplot.

@bfinzer Thanks. I believe the latest commit fixes all the above except for the equation jump on drag. I'd like to work on that separately since there's a lot going on in this PR already and I believe that problem exists in at least one other adornment  (Movable Line) as well.

Could you please confirm that the other bugs are fixed now?

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:
```diff
@@ -0,0 +1,198 @@
+}))
+.views(self => ({
+  get isValid() {
+    return self.intercept && self.rSquared && self.slope && self.sdResiduals
```
You probably want `isFinite()` here because this will fail when any of the values `=== 0`.

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-component.tsx:
```diff
@@ -122,7 +122,7 @@ export const LSRLAdornment = observer(function LSRLAdornment(props: IProps) {
 
       if (isFinished) {
-        const lines = model.lines.get(instanceKey)
+        const lines = dataConfig && model.getLines(xAttrId, yAttrId, cellKey, dataConfig)
```
Given that this occurs multiple times in multiple other callbacks, you may want to pull it out into its own callback:
```
  const getLines = useCallback(() => {
    return dataConfig && model.getLines(xAttrId, yAttrId, cellKey, dataConfig)
  }, [cellKey, dataConfig, xAttrId, yAttrId])
```
which then shortens the dependency list for the other effects.

kswenson commented on v3/src/components/graph/adornments/lsrl/lsrl-adornment-model.ts:
```diff
@@ -50,30 +61,29 @@ export const LSRLAdornmentModel = AdornmentModel
-      (!mse && mse !== 0) ||
-      (!xMean && xMean !== 0) ||
-      (!xSumSquaredDeviations && xSumSquaredDeviations !== 0)
+      !line || !line.intercept || !line.slope || count == null || mse == null || xMean == null ||
```
`line.intercept` and `line.slope` can validly be 0, so the test should be `== null` for them as well.

kswenson commented on v3/src/components/graph/utilities/graph-utils.ts:
```diff
@@ -297,28 +299,30 @@ export function computeSlopeAndIntercept(xAxis?: IAxisModel, yAxis?: IAxisModel)
+
+   // Under certain circumstances (adding new case) an empty value can sneak in here. Filter out.
+   iCoordPairs = iCoordPairs.filter((iPair: Point) => {
+     return !(!iPair.x || !iPair.y)
```
These should probably be `isFinite()` tests as well to handle 0 values.

kswenson reviewed (changes requested):
👍 Changes generally look good -- there's a few places where tests should be changed to properly handle 0 values.

bfinzer reviewed (approved):
👍🏻LGTM
I checked the bugs that I previously identified and they are all fixed except, as you noted, the equation jumping bug.
However, please note that the bug regarding the negative sign for the intercept also still occurs for a movable line equation.

emcelroy commented:
> However, please note that the bug regarding the negative sign for the intercept also still occurs for a movable line equation.

@bfinzer Thanks. I'll fix that in the locked intercept work.

kswenson reviewed (approved):
👍 Changes look good
